torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -6,16 +6,171 @@ namespace rpc {
 
 constexpr size_t WorkerInfo::MAX_NAME_LEN;
 
+// kRpcRetryDuration is the initial duration between consecutive RPC retries.
+constexpr auto kRpcRetryDuration = std::chrono::milliseconds(1000);
+// kBackoff is the multiplier for the exponential backoff algorithm.
+constexpr float kBackoff = 1.5;
+
 RpcAgent::RpcAgent(
     WorkerInfo workerId,
     std::unique_ptr<RequestCallback> cb,
     std::chrono::milliseconds rpcTimeout)
     : workerInfo_(std::move(workerId)),
       cb_(std::move(cb)),
       rpcTimeout_(rpcTimeout),
-      profilingEnabled_(false) {}
+      profilingEnabled_(false) {
+  rpcAgentRunning_.store(true);
+  rpcRetryThread_ = std::thread(&RpcAgent::retryExpiredRpcs, this);
+}
+
+RpcAgent::~RpcAgent() {
+  if (rpcAgentRunning_.load()) {
+    shutdown();
+  }
+}
+
+void RpcAgent::shutdown() {
+  if (!rpcAgentRunning_.exchange(false)) {
+    return;
+  }
+  rpcRetryMapCV_.notify_one();
+  rpcRetryThread_.join();
+  shutdownImpl();
+}
+
+std::shared_ptr<FutureMessage> RpcAgent::sendWithRetries(
+    const WorkerInfo& to,
+    Message&& message,
+    int maxRetries) {
+  auto future = std::make_shared<FutureMessage>();
+  steady_clock_time_point newTime = computeNewRpcRetryTime(/* retryCount */ 0);
+  Message& msg = message;
+  auto fm = send(to, std::move(message));
+
+  fm->addCallback([this, future, newTime, &to, &msg, maxRetries](
+                      const rpc::Message& lambdaMessage,
+                      const c10::optional<utils::FutureError>& futErr) {
+    if (futErr) {
+      LOG(INFO) << "Send try 1 failed\n";
+      if (maxRetries > 1) {
+        {
+          std::lock_guard<std::mutex> lock(rpcRetryMutex_);
+          rpcRetryMap_[newTime].emplace(std::make_shared<RpcRetryInfo>(
+              to, std::move(msg), future, 1, maxRetries));
+        }
+        rpcRetryMapCV_.notify_one();
+      } else {
+        std::string errorMessage;
+        errorMessage
+            .append(
+                "The RPC has not succeeded after the specified number of max retries (")
+            .append(std::to_string(1))
+            .append(").\n");
+        future->setError(errorMessage);
+      }
+    } else if (!future->completed()) {
```

Comment:
Why do we need this check? Is it possible the future has already been marked as completed by some other mechanism?

Reply:
If another send attempt was completed successfully and already marked the future as completed, then we shouldn't do anything. The `future::markCompleted()` has a TORCH_CHECK that throws if the future has already been completed, so this will guard against that. I believe @mrshenli suggested this might be a good check in earlier review.

Reply:
If another send attempt marked the future as completed successfully, then we wouldn't continue retries and wouldn't reach here. IIUC, it looks like the future can be marked completed only once since we only have one outstanding send at a time. Once a send() function completes (successfully/unsuccessfully), only then we kick off a retry.

Reply:
Ah yes you're right - the previous case was only possible back when the map wasn't manipulated through the callback. I'll change this per your suggestion.